File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution(object):
    def minWindow(self, s, t):
        if len(s) < len(t):
            return ""
        map = {}
        for char in t:
            if char in map:
                map[char] += 1
            else:
                map[char] = 1
        count = 0
        start = 0
        min_length = float("inf")
        min_start = 0
        for end in range(len(s)):
            if s[end] in map:
                map[s[end]] -= 1
                if map[s[end]] >= 0:
                    count += 1
            while count == len(t):
                if min_length > end - start + 1:
                    min_length = end - start + 1
                    min_start = start
                if s[start] in map:
                    map[s[start]] += 1
                    if map[s[start]] > 0:
                        count -= 1
                start += 1
        return "" if min_length == float("inf") else s[min_start:min_start+min_length]
                

File: 0076-minimum-window-substring/0076-minimum-window-substring.kt
class Solution {
    fun minWindow(s: String, t: String): String {
        val freq = IntArray(128) { 0 }
        t.forEach { freq[it.toInt()]++ }

        var res = ""
        var count = t.length
        var l = 0
        var r = 0

        while (r < s.length) {
            if (freq[s[r].toInt()]-- > 0) count--

            while (count == 0) {
                if (res.isEmpty() || r - l + 1 < res.length) {
                    res = s.substring(l, r + 1)
                }
                freq[s[l].toInt()]++
                if (freq[s[l++].toInt()] > 0) count++
            }
            r++
        }
        return res
    }
}
